File: Ahorn/entities/fieldGravityController.jl
module GravityHelperFieldGravityController

using ..Ahorn, Maple

@mapdef Entity "GravityHelper/FieldGravityController" FieldGravityController(
    x::Integer, y::Integer,
    persistent::Bool=false,
    arrowOpacity::Real=0.5,
    fieldOpacity::Real=0.15,
    particleOpacity::Real=0.5
)

const placements = Ahorn.PlacementDict(
    "Field Gravity Controller (GravityHelper)" => Ahorn.EntityPlacement(
        FieldGravityController,
    ),
)

const sprite = "objects/GravityHelper/gravityController/circle"
const sprite_dot = "objects/GravityHelper/gravityController/circle_dot"
const sprite_field = "objects/GravityHelper/gravityController/field"

function Ahorn.selection(entity::FieldGravityController)
    x, y = Ahorn.position(entity)
    return Ahorn.getSpriteRectangle(sprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::FieldGravityController)
    icon = get(entity.data, "persistent", false) ? sprite_dot : sprite
    Ahorn.drawSprite(ctx, icon, 0, 0)
    Ahorn.drawSprite(ctx, sprite_field, 0, 0)
end

end

File: Loenn/entities/fieldGravityController.lua
local consts = require("mods").requireFromPlugin("consts")
local helpers = require("mods").requireFromPlugin("helpers")
local drawableSprite = require("structs.drawable_sprite")

local placementData = helpers.createPlacementData('1', {
    persistent = false,
    arrowOpacity = 0.5,
    fieldOpacity = 0.15,
    particleOpacity = 0.5,
})

local fieldGravityController = {
    name = "GravityHelper/FieldGravityController",
    depth = -8500,
    ignoredFields = consts.ignoredFields,
    placements = {
        {
            name = "normal",
            data = helpers.union(placementData),
        },
    },
}

function fieldGravityController.sprite(room, entity)
    local spriteName = entity.persistent and "objects/GravityHelper/gravityController/icon_dot" or "objects/GravityHelper/gravityController/icon"
    local iconSprite = drawableSprite.fromTexture(spriteName, entity)
    local typeSprite = drawableSprite.fromTexture("objects/GravityHelper/gravityController/field", entity)
    return {iconSprite, typeSprite}
end

return fieldGravityController
